src/models/Ruleserver.ts

```diff
@@ -0,0 +1,290 @@
+        }
+        this.rulesByToken[eventRules.token].push(eventRules);
+    }
+
+    /**
+     * Stop serving the rules from this policy rule.
+     * @param eventRules The EventRules to stop serving from the ruleserver.
+     */
+    private stopEventRules(eventRules: EventRules): void {
+        const eventId = eventRules.eventId;
+        this.rulesByEvent.get(eventRules.roomId)?.delete(eventId);
+        const index = this.rulesByToken[eventRules.token].indexOf(eventRules);
+        if (index > -1) {
+            this.rulesByToken[eventRules.token].splice(index, 1);
+        }
+        eventRules.ruleserverRules.map(rule => this.stoppedRules[this.currentToken].push(rule.id));
+    }
+
+    /**
+     * Update the ruleserver to reflect the ListRule change.
+     * @param change A ListRuleChange sourced from a BanList.
+     */
+    private applyRuleChange(change: ListRuleChange): void {
+        if (change.changeType === ChangeType.Added) {
+            const eventRules = new EventRules(change.event.event_id, change.event.room_id, toRuleServerFormat(change.rule), this.currentToken);
+            this.addEventRules(eventRules);
+        } else if (change.changeType === ChangeType.Modified) {
+            const entry: EventRules|undefined = this.getEventRules(change.event.roomId, change.previousState.event_id);
+            if (entry === undefined) {
+                LogService.error('RuleServer', `Could not find the rules for the previous modified state ${change.event['state_type']} ${change.event['state_key']} ${change.previousState?.event_id}`);
+                return;
+            }
+            this.stopEventRules(entry);
+            const eventRules = new EventRules(change.event.event_id, change.event.room_id, toRuleServerFormat(change.rule), this.currentToken);
+            this.addEventRules(eventRules);
+        } else if (change.changeType === ChangeType.Removed) {
+            // When a redaction has happened, the unredacted version of the redacted event is in the previousState, the same event
+            // will be in the event slot (of RuleChange) with the redaction applied.
+            // When an event has been "soft redacted" (a new event with the same state keys with no content),
+            // the events in the previousState and event slots will be distinct events.
+            // In either case we can therefore use previousState to get the right event id to stop.
+            const entry: EventRules|undefined = this.getEventRules(change.event.room_id, change.previousState.event_id);
+            if (entry === undefined) {
+                LogService.error('RuleServer', `Could not find the rules for the previous modified state ${change.event['state_type']} ${change.event['state_key']} ${change.previousState?.event_id}`);
+                return;
+            }
+            this.stopEventRules(entry);
+        }
+    }
+
+    /**
+     * Watch the ban list for changes and serve its policies as rules.
```

Comment:
If I understand correctly, this does not serve any policies that have already been synced from the server, right?

Could you document this?

Reply:
Have changed it, but if we really needed a strong guarantee we could change BanList to call the listener with "all the changes up until now" when it is added or something like that.